sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.scripting
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.trees.{Origin, WithOrigin}
+
+/**
+ * Trait for all SQL scripting execution nodes used during interpretation phase.
+ */
+sealed trait CompoundStatementExec extends Logging {
+  /**
+   * Whether the statement originates from the SQL script or is created during the interpretation.
+   * Example: DropVariable statements are automatically created at the end of each compound.
+   */
+  val isInternal: Boolean = false
+
+  /**
+   * Reset execution of the current node.
+   */
+  def reset(): Unit
+}
+
+/**
+ * Leaf node in the execution tree.
+ */
+trait LeafStatementExec extends CompoundStatementExec
+
+/**
+ * Non-leaf node in the execution tree.
+ * It is an iterator over executable child nodes.
+ */
+trait NonLeafStatementExec extends CompoundStatementExec with Iterator[CompoundStatementExec]
+
+/**
+ * Executable node for SingleStatement.
+ * @param parsedPlan Logical plan of the parsed statement.
+ * @param origin Origin descriptor for the statement.
+ * @param isInternal Whether the statement originates from the SQL script or it is created during
+ *                   the interpretation. Example: DropVariable statements are automatically created
+ *                   at the end of each compound.
+ */
+class SingleStatementExec(
+    var parsedPlan: LogicalPlan,
+    override val origin: Origin,
+    override val isInternal : Boolean)
+  extends LeafStatementExec
+  with WithOrigin {
+
+  /**
+   * Whether this statement had to be executed during the interpretation phase.
+   * Example: Statements in conditions of If/Else, While, etc.
```

Comment:
Do we have test cases that include If/Else and While statements? Are they already supported, or will they be supported in the future?

Reply:
These statements will follow-up in the upcoming PRs, we already have them ready/tried out in our POC. I can share more details offline if you are interested.

For now, we are introducing just the basic structure in order to have simpler PRs for reviewing.